2022-11-23_proposal-process.md

```diff
@@ -0,0 +1,129 @@
+
+Given initial feedback, I would propose to not render proposals on Prometheus website because it can cause confusion of proposed features being in stable codebase and treating those design documents as feature documentation.
+
+One good argument against repositories like this or websites with design docs is the beauty of stable URL of the Google Document, so it can be referenced and valid longer. I propose treating the GitHub repo name and file name in very stable way and (ideally) never change it after acceptance. This will result in similar, stable characteristics and ability to get `latest` version of same proposal or even referencing permanent link to exact version of the design document.
+
+### Details
+
+I propose to host in git only accepted proposals, simply in the root directory. It will roughly look like this:
+
+![repo view.png](assets/proposal-repo.png)
+
+Given that, the process of proposing change with the design doc would look as follows (we can use below text as the initial instruction):
+
+1. Fork `github.com/prometheus/proposals`
+2. Create a GitHub Pull Request with a design document in markdown format to root directory of the repository. Make sure to use [template](0000-00-00_template.md) the guide for what sections should be present in the document. Put creation date (the day you started preparing this design doc) as the prefix and some unique name as the suffix in the file name.
+   1. If you prefer Google Docs on any other collaboration tool, feel free to use it in the initial state. we recommend [Open Source Design Doc Template](https://docs.google.com/document/d/1zeElxolajNyGUB8J6aDXwxngHynh4iOuEzy3ylLc72U/edit#). However, the approval process will only happen officially in the Pull Request.
+3. Automatic formatter is enabled in the repository. Use `make` locally to format it. Use `make check` to check all link (will be done on the CI too).
+4. The design is accepted if the PR is merged to this repository. It's totally ok to decide eventually to reject the proposal and simply close the PR with meaningful reasons of why it was rejected.
+   1. If more eyes are needed or no consensus was made: Propose and announce your idea on
+      [Prometheus DevSummit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit)
+      or mailing list to gather more information. You are welcome to start working on design doc before bigger discussion--it is often easier
+      to start discussion with prior information prepared. Be prepared that the idea might be rejected later--still the record of the document in the Pull Request is useful even in rejected state to inform about past decisions and opportunities considered.
+   2. Optionally: Find sponsor among Prometheus maintainers to get momentum on a change.
```

Comment:
IMHO this is probably the main pain point. I'm not a maintainer and just someone from the peanut gallery so maybe I'm wrong but I've seen so many good ideas proposed during dev summits but most of them never go anywhere. How to find a sponsor? For example, https://github.com/prometheus/prometheus/issues/3806 is almost 5 years old but there's still no rate() function in upstream Prometheus that doesn't do extrapolation. :/ I am personally interested in this because it's hard to explain to others how come rate() gives this or another result. xrate is much nicer. I even updated the code to work with a newer version of Prometheus but I have no clue how to pursue that further. It would be nice to include something about this.

Reply:
Great point. I think it will be reconsidered if you would propose it again. I think this proposal process will only help with those problems, don't you think?

There will always be contentious ideas and proposals, but I think Prometheus acknowledged within last year that we need to be more permissive with new ideas, thus my proposal (:

Reply:
I think we should primarily focus on solving the typical problems that happen when a change or a new feature is proposed.

The typical fate of an idea discussed at a dev summit is the following:
* Generally, the idea is liked and looks promising.
* The dev summit decides that a design doc is needed.
* More discussions will happen once a design doc exists.
* Nobody steps up to write the design doc.

There is actually no problem with not being permissive enough. The problem is that nobody is doing the work of writing the design doc. The proposal process here intends to unify and facilitate the creation of design docs, and that might actually help with the typical fate described above.

While the project had gained a reputation of "saying no" in the past, I think it has become much more open to change in recent years. There are IMHO very few proposed changes that have ended up in a controversial rejection. Most are indeed in the limbo described above: More research needed, but nobody steps up to do it.

The discussion around `rate` is a notable exception here, but I think it's structurally so different that it shouldn't be the prime target for the unified proposal process. In contrast to most other proposals, it has actually seen a lot of discussion. A design doc already exists. The problem here is not that the discussion is ignored by the Prometheus maintainers (even if it might appear that way from the outside), the problem is that not a single one is convinced enough about the benefits to champion it. I should also note that the existing implementation is already based on long discussions and extensive research, which is again rather special, as most other proposals are either about a completely new feature or at least intend to change something that is less well explored and established.